clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -6082,6 +6082,22 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, const TypeSourceI
       Self.Diag(Rhs->getTypeLoc().getBeginLoc(), diag::err_vla_unsupported)
           << 1 << tok::kw___is_layout_compatible;
     return Self.IsLayoutCompatible(LhsT, RhsT);
+  }
+  case BTT_IsPointerInterconvertibleBaseOf: {
+    if (!LhsT->isUnionType() && !RhsT->isUnionType() &&
+        !Self.getASTContext().hasSameUnqualifiedType(LhsT, RhsT)) {
+      Self.RequireCompleteType(Rhs->getTypeLoc().getBeginLoc(), RhsT,
+                               diag::err_incomplete_type);
+    }
+
+    if (LhsT->isVariableArrayType())
+      Self.Diag(Lhs->getTypeLoc().getBeginLoc(), diag::err_vla_unsupported)
+          << 1 << tok::kw___is_pointer_interconvertible_base_of;
+    if (RhsT->isVariableArrayType())
+      Self.Diag(Rhs->getTypeLoc().getBeginLoc(), diag::err_vla_unsupported)
+          << 1 << tok::kw___is_pointer_interconvertible_base_of;
```

Comment:
> Do we need these checks? Is there a reason we can’t just ‘fall through’ here if the types are VLAs?

My perspective is "we need less VLA extensions", and from this point of view those checks are beneficial.

> but in this case this is about base classes, and since a VLA isn’t a class type to begin with, just returning false would make sense imo.

This type trait accepts a broader set of types than just class types, so it's not just about types that might participate in inheritance.

CC @erichkeane

Reply:
I'm happy with not supporting VLA here.
Should we factor out a `DiagnoseVLAInCXXTypeTrait`?

Reply:
> Should we factor out a DiagnoseVLAInCXXTypeTrait?

Makes sense. I'll do that and apply it across type traits that already check for VLAs in a subsequent NFC patch.